summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala

```diff
@@ -107,13 +136,15 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
     if (pendingFuturesCount > maxWaitingFuturesCount) {
       // Too many futures waiting, let's clear.
       outstandingFutures.dequeueAll(_.isDefined)
```

Comment:
We might keep hitting the limit in terms of pending futures too. We get events in the bolt at a great rate, immediately after we clear the futures to get down to the limit we'll get another event and check again, by that time, or a few events later, we might again have more pending futures than the limit. Clearing defined futures will not help because the pending futures themselves may go over the limit. Having a lower watermark will avoid that.

I don't have a benchmark suite. I do have a job, profiling which I found the original issue. I guess it would be hard to see the effect of just the watermark change. But overall I do see improvement with all the fixes in this PR so far. I'm taking the yourkit profile to monitor how long the slave thread blocks waiting for the futures to clear. That period is certainly down, so the slave thread is consuming events at a greater rate.